enhancements/testing/openshift-tests-extension.md

```diff
@@ -0,0 +1,646 @@
+execution. It includes information like the random seed being used for test execution
+planning as the number of times a test has been run against a cluster.
+
+Rarely, Component Readiness may analyze results using these dimensions to, for example,
+determine that different runs of the same test on the same cluster behave differently
+in a statistically significant way.
+
+### Test Extension Binaries
+Test extension binaries will be defined as a first class method of introducing new tests and suites
+into OpenShift origin's test framework. Historically, all tests had to be contributed to 
+github.com/openshift/origin, but extension binaries can be developed in other repositories
+and built independently of origin.
+
+Extension binaries must implement an "extension interface" consisting of CLI verbs
+and arguments that allow tests to be discovered and executed. The extension interface
+is non-trivial to implement, so to simplify the creation of extensions, a Go module & 
+repository will be maintained by TRT: https://github.com/openshift-eng/openshift-tests-extension . 
+
+Vendoring this module and following its integration guide will provide component test authors
+the majority of the logic necessary to fulfill the test extension interface. It will
+also provide a means of centrally defining new testing requirements. As component owners re-vendor
+the module, they would be required to implement those new testing requirements
+for the module to compile and run a test extension binary successfully.
+
+During a run of the origin test framework (`openshift-tests`), the framework will:
+1. Discover available extension binaries.
+2. Discover metadata and component configuration options for those extensions (running `info` verb and parsing standard out).
+3. Discover tests available from those extensions (running `list` verb and parsing standard out).
+4. Plan an appropriate and efficient testing strategy based on the framework invocation.
+5. Execute the tests using subsequent invocations of the discovered binaries (`run-test` invocations).
+6. Collecting the result and output of the test invocations and aggregating it into overall test suite results.
+
+
+#### Binary Discovery
+The existence of test extension binaries can be registered one of two ways by test authors. 
+
+##### OpenShift Payload Extension Binaries
+For OpenShift payload components contributors can advertise the existence of an extension binary
+by adding information (the imagestream tag for the OCP payload component and the path to the binary 
+within their image) to a simple registry datastructure in github.com/openshift/origin. 
+
+##### Non-Payload Extension Binaries
+For non-payload components, contributors must advertise their extension binary for `openshift-tests`
+to discover. This is accomplished by creating an ImageStream / ImageStreamTag with a special label:
+
+`testextension.redhat.io/component=<product-name>-<component-name>`

```

Comment:
I think you are talking about image labels here? Which on tagging get transcribed to an ImageStreamTag annotation? Are we talking about an imagestream populated on the CI cluster, or one included in the payload manifests?

Couple potential issues there:
1. ART ImageStreams don't tag non-payload images
2. Not all payloads are built from imagestreams... multi for instance is not. I suppose these could get the info directly from the image labels?

Reply:
I'm describing an ImageStream instantiated on the cluster under test / ephemeral cluster. The flow would be:
1. A CI job stands up an cluster to test.
2. The job instantiates a `TestExtensionAdmission` on the ephemeral cluster.
3. The job installs an optional operator that wants to participate in testing.
4. The optional operator (or CI job) creates an ImageStream and labels it according the described labeling pattern to advertise a tests extension binary.
5. openshift-tests runs, analyzes the cluster under test for ImageStreams advertising extension binaries, verifies they are permitted by `TestExtensionAdmission` instance(s) and extracts them.